### PR TITLE
schedule: respect limits in concurrent patrol workers

### LINKED ISSUE #10926: patrol-region-worker-count can exceed schedule limits
## Bug Report

When `schedule.patrol-region-worker-count` is greater than 1, multiple patrol workers can check the current operator count concurrently and all generate operators before any worker adds its operator to the operator controller.

As a result, patrol region checking can exceed schedule limits such as `replica-schedule-limit`, and similar races can affect merge or affinity operators.

## Expected Behavior

Concurrent patrol workers should respect the configured schedule limits.

## Affected Area

- `pkg/schedule/checker.Controller.PatrolRegions`
- operator admission from patrol region workers

## Related PR

Fix: https://github.com/tikv/pd/pull/10925

### PR DESCRIPTION
**Pull request by rleungx**
### What problem does this PR solve?

Issue Number: close #10926

When `schedule.patrol-region-worker-count` is greater than 1, multiple patrol workers can check operator counts concurrently and generate operators before any worker adds its operator to the operator controller. This can make patrol region checking exceed schedule limits such as `replica-schedule-limit`.

### What is changed and how does it work?

```commit-message
schedule: respect limits in concurrent patrol workers
```

This PR:

- serializes the patrol-worker and priority-region operator admission paths with one controller mutex
- keeps the existing checker flow and schedule-limit checks inside `CheckRegion`, then enqueues the generated operators while still holding the same admission mutex
- adds a regression test for concurrent patrol workers respecting `replica-schedule-limit`

### Check List

Tests

- Unit test

```bash
go test ./pkg/schedule/checker -count=1
go test ./server/cluster -run '^TestPatrolRegionConcurrencyRespectReplicaLimit$' -count=1 -timeout=2m
```

Code changes

- No configuration change
- No API change
- No persistent data change

Side effects

- No user-facing side effects

Related changes

- Need to cherry-pick to the release branch

### Release note

```release-note
Fix an issue where concurrent patrol region workers could exceed schedule limits.
```

### CODE REVIEW

File: pkg/schedule/checker/checker_controller.go [hunk 1]
```diff
@@ -147,20 +151,14 @@ func (c *Controller) PatrolRegions() {
...
-			if c.cluster.IsSchedulingHalted() {
-				for len(c.patrolRegionContext.regionChan) > 0 {
-					<-c.patrolRegionContext.regionChan
-				}
-				log.Debug("skip patrol regions due to scheduling is halted")
+			// wait for the previous batch to be fully processed.
+			if !c.patrolRegionContext.isIdle() {
 				continue
 			}
 			c.metrics.patrolRegionChannelSize.Set(float64(len(c.patrolRegionContext.regionChan)))
```
Comment (lhy1024) [inline pkg/schedule/checker/checker_controller.go]: This metric now effectively only records 0. The code checks `isIdle()` before setting `patrolRegionChannelSize`, so by the time we reach this line there should be no queued or in-flight patrol work, and `len(regionChan)` no longer reflects patrol backlog.

Could we either sample before the idle check, or change the metric to report the pending/in-flight count instead? `phaseWaitForChannel` also no longer gets observed after this change, so it should either be restored or removed to avoid misleading metrics.

File: tests/server/apiv2/handlers/affinity_test.go [hunk 2]
```diff
@@ -139,6 +139,42 @@ func mustGetAllAffinityGroups(re *require.Assertions, serverAddr string) *handle
 	return &result
 }
 
+func waitForAffinityGroups(re *require.Assertions, url string, groupIDs ...string) {
```
Comment (lhy1024) [inline tests/server/apiv2/handlers/affinity_test.go]: Could we drop the changes in this file, or rebase and keep the current master version? This PR now has a content conflict with latest master in `tests/server/apiv2/handlers/affinity_test.go`, and master already contains an equivalent `tryGetAffinityGroups` / `waitForAffinityGroups` polling helper.

The affinity test stabilization is also unrelated to the patrol-worker schedule-limit fix, so keeping it in this PR adds conflict surface and review noise.

File: pkg/schedule/checker/checker_controller.go [hunk 3]
```diff
@@ -147,22 +151,17 @@ func (c *Controller) PatrolRegions() {
 				failpoint.Continue()
 			})
 			c.updateTickerIfNeeded(ticker)
+			pendingRegionCount := c.patrolRegionContext.pendingRegionCount()
+			c.metrics.patrolRegionChannelSize.Set(float64(pendingRegionCount))
```
Comment (lhy1024) [inline pkg/schedule/checker/checker_controller.go]: This gauge now reports the pending patrol work count (queued plus in-flight regions), not the raw channel length anymore. Could we update the metric help/comment in `metrics.go` to match the new semantics? Keeping the existing metric name for compatibility seems fine, but the current help text still says "Size of patrol region channel", which is no longer accurate.

**Review by lhy1024 — COMMENTED** [review]
the rest LGTM

File: pkg/schedule/checker/checker_controller.go [hunk 4]
```diff
@@ -237,6 +261,41 @@ func (c *Controller) updatePatrolWorkersIfNeeded() {
...
 
+// operatorLimit is rechecked while admitting a candidate into the operator queue.
+type operatorLimit struct {
+	kind        operator.OpKind
+	checkerType types.CheckerSchedulerType
+	getLimit    func() uint64
+}
+
+func (l operatorLimit) reached(opController *operator.Controller) bool {
+	return l.getLimit != nil && opController.OperatorCount(l.kind) >= l.getLimit()
```
Comment (lhy1024) [inline pkg/schedule/checker/checker_controller.go]: I think the schedule-limit recheck is still incomplete for merge / affinity-merge candidates.

`operatorLimit.reached()` only checks `OperatorCount(kind) >= limit` before admission, but a candidate can contain multiple operators. Both `MergeChecker.Check()` and affinity merge return a pair, and `AddWaitingOperator()` can admit both operators together. For example, with `merge-schedule-limit = 3` and current `OperatorCount(OpMerge) = 2`, this check still passes, then the next merge pair is admitted and the final count becomes 4.

This seems within this PR's stated scope because the PR says it re-checks replica, merge, and affinity schedule limits immediately before `AddWaitingOperator`.

Could we either account for the candidate size when checking the limit, e.g. reject `current + candidateCount > limit`, or clarify that merge / affinity-merge limits may still exceed the configured limit by one pair? It would also be good to add a regression test for this boundary case.

File: pkg/schedule/checker/checker_controller.go [hunk 5]
```diff
@@ -403,6 +413,15 @@ func (c *Controller) tryAddOperators(region *core.RegionInfo) {
 		c.RemovePendingProcessedRegion(id)
 		return
 	}
+	c.operatorMu.Lock()
```
Comment (lhy1024) [inline pkg/schedule/checker/checker_controller.go:428]: The new coarse-grained `operatorMu` does fix the stale-count race between patrol workers, but I think the merge / affinity-merge boundary case is still not covered.

`CheckRegion()` still only checks `OperatorCount(OpMerge) < GetMergeScheduleLimit()` before returning a merge pair. Since a merge candidate contains two operators, `merge-schedule-limit = 3` and current `OperatorCount(OpMerge) = 2` can still admit another pair and end up with 4 merge operators.

I verified this on the latest head with the following focused test, which fails because the final merge count becomes 4:

```go
package checker

import (
	"context"
	"testing"

	"github.com/stretchr/testify/require"

	"github.com/tikv/pd/pkg/mock/mockcluster"
	"github.com/tikv/pd/pkg/mock/mockconfig"
	"github.com/tikv/pd/pkg/schedule/hbstream"
	"github.com/tikv/pd/pkg/schedule/operator"
)

func TestTryAddOperatorsRejectsMergePairPastLimit(t *testing.T) {
	re := require.New(t)
	ctx, cancel := context.WithCancel(context.Background())
	defer cancel()

	opt := mockconfig.NewTestOptions()
	tc := mockcluster.NewCluster(ctx, opt)
	tc.SetMergeScheduleLimit(3)
	tc.SetSplitMergeInterval(0)
	for storeID := uint64(1); storeID <= 3; storeID++ {
		tc.AddRegionStore(storeID, 0)
	}
	tc.AddLeaderRegionWithRange(1, "", "a", 1, 2, 3)
	tc.AddLeaderRegionWithRange(2, "a", "b", 1, 2, 3)
	tc.AddLeaderRegionWithRange(3, "b", "c", 1, 2, 3)
	tc.AddLeaderRegionWithRange(4, "c", "d", 1, 2, 3)

	stream := hbstream.NewTestHeartbeatStreams(ctx, tc, false)
	defer stream.Close()
	oc := operator.NewController(ctx, tc.GetBasicCluster(), tc.GetSharedConfig(), stream)
	controller := NewController(ctx, tc, tc.GetCheckerConfig(), oc)

	controller.tryAddOperators(tc.GetRegion(1))
	re.Equal(uint64(2), oc.OperatorCount(operator.OpMerge))

	controller.tryAddOperators(tc.GetRegion(3))
	re.LessOrEqual(oc.OperatorCount(operator.OpMerge), uint64(3))
}
```

Could we either account for the candidate size when checking merge / affinity limits, or explicitly document that these limits may be exceeded by one pair?

Comment (rleungx) [reply pkg/schedule/checker/checker_controller.go:428]: Fixed in 8e7ec8085. Merge and affinity candidates now check their full candidate count against the remaining schedule-limit capacity (`candidateCount <= limit - current`) before admission, so a pair is rejected when only one slot remains. I added the provided merge boundary regression plus focused capacity cases. Verified with the full checker package, the existing patrol replica-limit regression, focused race tests, and `make static`.

File: pkg/schedule/checker/checker_controller.go (hunk 4 above)
Comment (rleungx) [reply pkg/schedule/checker/checker_controller.go]: Addressed in 8e7ec8085 with candidate-count-aware capacity checks for both merge and affinity candidates, plus the odd-limit merge regression test.

File: pkg/schedule/checker/checker_controller.go (hunk 5 above)
Comment (lhy1024) [reply pkg/schedule/checker/checker_controller.go:428]: I want to clarify my earlier suggestion here. The concern about merge candidates being different from single-operator candidates is still valid, but suggesting a strict `current + candidateCount <= limit` check was too strict if we want to preserve the existing merge-pair semantics.

The original issue this PR is fixing is the stale-count race between concurrent patrol workers. It is not necessarily trying to change the existing single-threaded behavior where one merge schedule produces two operators. Existing behavior/tests seem to allow one merge pair to be created when `merge-schedule-limit = 1`, and existing `TestPatrolRegionConcurrency` uses `mergeScheduleLimit = 15` and waits for at least 15 operators.

With the current `len(ops)`-based capacity check, an odd merge limit can stop below the configured limit. For example, with `merge-schedule-limit = 15`, the merge checker can stop at 14 because the next merge pair has 2 operators. I verified this on the latest head by running:

```bash
make gotest GOTEST_ARGS='./server/cluster -run TestPatrolRegionConcurrency -count=1 -timeout=2m'
```

and `TestPatrolRegionConcurrency` fails because the `len(oc.GetOperators()) >= mergeScheduleLimit` condition is never satisfied.

Could we preserve the existing merge-pair semantics while still serializing admission to avoid the concurrent stale-count overshoot? In other words, the admission lock should prevent multiple workers from racing on the same old count, but we probably should not make odd merge limits reject a whole pair before reaching the configured limit.
